src/data/setoid.lean

```diff
@@ -0,0 +1,603 @@
+
+/-- A partition of α does not contain the empty set. -/
+lemma ne_empty_of_mem_partition {c : subtype (partitions α)} {s} (h : s ∈ c.1) : 
+  s ≠ ∅ :=
+λ hs0, c.2.1 $ hs0 ▸ h 
+
+/-- All elements of a partition of α are the equivalence class of some y ∈ α. -/
+lemma exists_of_mem_partition {c : subtype (partitions α)} {s} (hs : s ∈ c.1) : 
+  ∃ y, s = { x | rel (mk_classes c.1 c.2.2) x y } :=
+let ⟨y, hy⟩ := set.exists_mem_of_ne_empty $ ne_empty_of_mem_partition hs in 
+  ⟨y, eq_eqv_class_of_mem c.2.2 hs hy⟩
+
+/-- The equivalence classes of the equivalence relation defined by a partition of α equal
+    the original partition. -/
+theorem classes_mk_classes (c : subtype (partitions α)) : 
+  (mk_classes c.1 c.2.2).classes = c.1 := 
+set.ext $ λ s,
+  ⟨λ ⟨y, hm, hs⟩, by rcases c.2.2 y with ⟨b, hc, hb, hy⟩; convert hc; rw hs; ext; 
+      exact ⟨λ hx, symm' (mk_classes c.1 c.2.2) hx b hc hb, λ hx b' hc' hx', 
+              eq_of_mem_eqv_class c.2.2 hc hx hc' hx' ▸ hb⟩, 
+   λ h, let ⟨y, hy⟩ := set.exists_mem_of_ne_empty $ ne_empty_of_mem_partition h in 
+     ⟨y, hy, eq_eqv_class_of_mem c.2.2 h hy⟩⟩
+
+/-- Defining ≤ on partitions as the ≤ defined on their induced equivalence relations. -/
+instance partition.le : has_le (subtype (partitions α)) :=
+⟨λ x y, mk_classes x.1 x.2.2 ≤ mk_classes y.1 y.2.2⟩
+
+/-- Defining a partial order on partitions as the partial order on their induced
+    equivalence relations. -/
+instance partition.partial_order : partial_order (subtype (partitions α)) :=
+{ le := (≤),
+  lt := λ x y, x ≤ y ∧ ¬y ≤ x,
+  le_refl := λ x, @le_refl (setoid α) _ _,
+  le_trans := λ _ _ _ h1 h2, @le_trans (setoid α) _ _ _ _ h1 h2,
+  lt_iff_le_not_le := λ x y, iff.rfl,
+  le_antisymm := λ x y hx hy, let h := @le_antisymm (setoid α) _ _ _ hx hy in by 
+    rw [subtype.ext, ←classes_mk_classes x, ←classes_mk_classes y, h] }
+
+variables (α)
+
+/-- The order-preserving bijection between equivalence relations and partitions of sets. -/
+def partition.order_iso : 
+  ((≤) : setoid α → setoid α → Prop) ≃o (@setoid.partition.partial_order α).le := 
+{ to_fun := λ r, ⟨r.classes, empty_not_mem_classes, classes_eqv_classes⟩,
+  inv_fun := λ x, mk_classes x.1 x.2.2,
+  left_inv := λ r, mk_classes_classes r,
+  right_inv := λ x, by rw [subtype.ext, ←classes_mk_classes x],
+  ord := λ x y, by conv {to_lhs, rw [←mk_classes_classes x, ←mk_classes_classes y]}; refl}
+
+variables {α}
+
+/-- A complete lattice instance for partitions; there is more infrastructure for the 
+    equivalent complete lattice on equivalence relations. -/
+instance partition.complete_lattice : 
+  _root_.lattice.complete_lattice (subtype (partitions α)) :=
+galois_insertion.lift_complete_lattice $ @order_iso.to_galois_insertion
+_ (subtype (partitions α)) _ (partial_order.to_preorder _) $ partition.order_iso α
+ 
+theorem partition.le_def (c d : subtype (partitions α)) : 
+  c ≤ d ↔ mk_classes c.1 c.2.2 ≤ mk_classes d.1 d.2.2 := 
+iff.rfl
+
+-- Bell numbers 
```

Comment:
Maybe move this section into its own file.

Reply:
Thank you for the review. Where would be appropriate? That section's so patchy at the moment - would it be better to take it out for now and then PR a file about Bell numbers that actually proves they are what they're supposed to be at some point?

Reply:
That sounds good--splitting PRs into smaller pieces is usually easier for reviewers as well.

Reply:
Noted, thank you! I should've checked what the protocol is for PR lengths of entirely new files, I'd probably have caught more errors before opening the PR if I'd started with a smaller chunk anyway.